File: src/ca/uqam/mgl7460/tp1/types/traitements/instances/InstanceTache.java
package ca.uqam.mgl7460.tp1.types.traitements.instances;

import ca.uqam.mgl7460.tp1.types.modeles.DemandePret;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionTache;

/**
 * le type InstanceTache représente l'exécution d'une <code>DefinitionTache</code> faisant partie
 * d'une modèle de processus (<code>DefinitionProcessus</code>), sur une <code>DemandePret</code>,
 * dans le cadre d'un processus englobant (<code>InstanceProcessus</code>)
 */
public interface InstanceTache {

    /**
     * la définition de tâche dont cette instance est une exécution particulière.
     * 
     * C'est la définition de tâche qui me permet de "aller chercher la fonction
     * à exécuter" (voir méthode <code>executer</code> plus bas)
     * @return
     */
    public DefinitionTache getDefinitionTache();

    /**
     * retourne l'état d'exécution de la tache. C'est l'un de PRET, ENCOURS et TERMINE
     * @return
     */
    public EtatTraitement getEtatInstanceTache();

    /**
     * Modifie l'état de traitement de la tâche
     * @param etat
     */
    public void setEtatInstanceTache(EtatTraitement etat);

    /**
     * retourne le processus englobant, i.e. la <code>InstanceProcessus</code>
     * dont je fais partie (c-a-d dont je suis l'une des étapes)
     * @return
     */
    public InstanceProcessus getProcessusEnglobant();

    /**
     * retourne la demande de prêt sur laquelle je vais exécuter le traitement
     * @return
     */
    public DemandePret getDemandePret();

    /**
     * modifie la demande de pret sur laquelle je (vais) opérer
     * @param demande
     */
    public void setDemandePret(DemandePret demande);

    /**
     * Cette méthode exécute l'instance de tâche.
     * 
     * Elle va chercher le <code>TraitementTache</code> disponible dans sa
     * <code>definbitionTache</code> et l'exécuter.
     * 
     * En fonction du résultat (true ou false), elle modifie:
     * 1)   le <code>ResultatTraitement</code> de la demande de prêt
     * 2)   l'état de traitement de l'instance de tache elle même
     * 
     * Et elle va signaler au processus englobant qu'elle a fini, en appelant la méthode 
     * <code>void signalerFinTache(InstanceTache instanceTache) throws ExceptionDefinitionProcessus</code>
     * 
     * @throws ExceptionDefinitionProcessus
     */
    public void executer() throws ExceptionDefinitionProcessus;

}

File: src_python/traitements/instances/instance_tache.py
from modeles.demande_pret import DemandePret
from modeles.resultat import Resultat
from traitements.instances.etat_traitement import EtatTraitement
from traitements.instances.exception_definition_processus import ExceptionDefinitionProcessus

class InstanceTache:
    def __init__(self, processus_englobant, definition_tache):
        self.processus_englobant = processus_englobant
        self.definition_tache = definition_tache
        self.demande_pret = processus_englobant.demande_pret
        self.etat_instance_tache = None

    
    @property
    def demande_pret(self):
        return self._demande_pret
    @demande_pret.setter
    def demande_pret(self, value: DemandePret):
        self._demande_pret = value
    
    @property
    def etat_instance_tache(self):
        return self._etat_instance_tache
    @etat_instance_tache.setter
    def etat_instance_tache(self, value: EtatTraitement):
        self._etat_instance_tache = value

    def executer(self):
        traitement_tache = self.definition_tache.traitementTache
        if traitement_tache is None:
            raise ExceptionDefinitionProcessus(self.processus_englobant, [self.definition_tache], "La tache n'a pas de traitement.")
        else:
            resultat = traitement_tache(self.demande_pret, self.processus_englobant.logger)
            resultat_traitement = self.demande_pret.resultat_traitement

            if (resultat_traitement is None):
                self.demande_pret.resultat_traitement = Resultat.ACCEPTEE if resultat else Resultat.REFUSEE

            if resultat == False:
                resultat_traitement.ajouter_message(f"La tache {self.definition_tache.nom} a échoué.")
                resultat_traitement.resultat = Resultat.REFUSEE
            
            self.etat_instance_tache = EtatTraitement.TERMINE
            self.processus_englobant.signaler_fin_tache(self)
            